### DISCUSSION TITLE
EP1220

### DISCUSSION THREAD

**Discussion by Don87** [TillFleisch/ESPHome-Philips-Smart-Coffee#67] [Support for EPxxxx]
Hello Till,

Is it possible to add the Philips EP1220?

**Comment by TillFleisch** [comment]
Probably yes. Based on what we've learned so far from other discussions and machines, the underlying protocol seems to be similar across different machines. Messages originating from the display seem to contain the model number in some way. (It may still be the case that the EP22xx command also work on the EP12xx.)


I don't have access to an EP1220 therefore I can't give this a try. 
As far as I can tell there don't seem to be any major differences between the EP2220 and EP1220. Are you aware of differences between the two machines?

**Comment by Don87** [comment]
Hello Till, thank you for your answer. Your project was supposed to be my biggest Esphome project, but it didn't work out. HA shows the programming, but unfortunately I can't turn anything on or off. I took the second board that you suggested as an alternative.
The soldering result looks good. Maybe I forgot something on the
![20240419_223830](https://github.com/TillFleisch/ESPHome-Philips-Smart-Coffee/assets/45289133/d5f7529f-233d-4120-acde-a6b395d1e77d)
 board?

**Comment by TillFleisch** [comment]
> but unfortunately I can't turn anything on or off

I assume by this you mean you cannot interact with the machine through Home Assistant? (can't turn on/off the machine/prepare a beverage) 
In that case, it's likely that the protocols used in the machines are different enough, such that things don't work.

Does the status sensor report the state of the machine correctly?

You can use the [ESPHome UART Debug functionality](https://esphome.io/components/uart.html#debugging) to observe what is happening on the bus.
When observing messages originating from the display (RX on the display uart in your config) you should see messages similar to the ones in [protocol.md](https://github.com/TillFleisch/ESPHome-Philips-Smart-Coffee/blob/main/protocol.md) and [commands.h](https://github.com/TillFleisch/ESPHome-Philips-Smart-Coffee/blob/main/components/philips_coffee_machine/commands.h).
It seems to be the case that these messages contain the model NR in bytes 4-6 (zero-indexed).

EP2220
`0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x02, 0x01, 0x00, 0x00, 0x1D, 0x3B`

EP32xx
`0xD5, 0x55, 0x00, 0x01, 0x03, 0x00, 0x0E, 0x01, 0x00, 0x00, 0x39, 0x39`

I suspect that these are different on your machine (EP1200), somewhere along the line of `0x01, 0x00, 0x0?`
Thus, the checksum of injected messages will be invalid, which your machine will then ignore.

Anyhow, you can use the UART debug function to verify this is the case. In a similar manner, the required command for the EP1200 can be retrieved this way.


> Maybe I forgot something on the board?  

Seems ok. Double-check your wiring and yaml pin configuration. You may also want to look at the documentation of that MOSFET to see if it's oriented correctly. Which MOSFET are you using? Some have too high gate-threshold-voltage which can cause problems.

**Comment by AndrewHath103** [comment]
If you haven't found the commands yet I just wanted to add that I've been working on using an EP1220 myself and have been able to find most of them. These are the ones i have found and believe are right. I've gotten all the buttons to work on the machine with the commands beside being able to power it on. When I have tried to find the command for power on the machine the only one it sends is what i currently have for powerOff.

//commands
byte powerOn[] =       {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x01, 0x00, 0x00, 0x1A, 0x3C}; //Wrong
byte requestInfo[] =    {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x00, 0x00, 0x00, 0x16, 0x31}; //MAYBE WHAT IS ALWAYS SENT
byte powerOff[] =       {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x01, 0x00, 0x00, 0x1A, 0x3C}; //Believe right
byte hotWater[] =       {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x04, 0x00, 0x00, 0x26, 0x06};
byte espresso[] =        {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x02, 0x00, 0x00, 0x0E, 0x2A};
byte coffee[] =            {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x08, 0x00, 0x00, 0x3E, 0x1B};
byte steam[] =            {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x10, 0x00, 0x00, 0x0E, 0x21};
byte coffeePulver[] =  {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x00, 0x02, 0x00, 0x0E, 0x28};
byte coffeeWater[] =  {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x00, 0x04, 0x00, 0x27, 0x02};
byte calcNclean[] =    {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x00, 0x20, 0x00, 0x2F, 0x30};
byte aquaClean[] =    {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x00, 0x10, 0x00, 0x0A, 0x31};
byte startPause[] =     {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x00, 0x00, 0x01, 0x1E, 0x35};

**Reply by AndrewHath103** [reply]
For reference when i was finding the commands i used an ESP8266 with this wiring and Arduino code: 

<img width="566" alt="Schematic" src="https://github.com/TillFleisch/ESPHome-Philips-Smart-Coffee/assets/157642892/a968b2f0-3445-4099-b84b-62a83c418f32">


#include <ESP8266WiFi.h>
#include <SoftwareSerial.h>

#if defined(ESP8266) && !defined(D5)
#define D5 (14)
#define D6 (12)
#define D7 (13)
#endif

SoftwareSerial swSer(D5, D6); // SoftwareSerial to read Display TX

void setup() {
  Serial.begin(115200); // Initialize serial for debugging
  swSer.begin(115200);  // Initialize SoftwareSerial

  // Initialize other components and settings here
}

void loop() {
  // Check for available data from the coffee machine
  while (swSer.available() > 0) {
    char receivedByte = swSer.read();
    Serial.print(receivedByte, HEX); // Print received byte in hexadecimal format
    Serial.print(" ");               // Add space for readability
  }

  // You can add other logic or processing here if needed
}

**Reply by Don87** [reply]
Thank you for your help. You're both great. Unfortunately, I'm still a newbie. I installed Home Assistant 3 weeks ago. I have to delve deeper into the matter to understand it. I've wrecked 2 esp on other projects and have to wait for my order. Then I will try with uart debug and insert the commands in the right place. To learn and understand the whole thing. It's just something different than adding an nspanel with code snippets. 😀

**Reply by TillFleisch** [reply]
On the `EP2220` only the power off `command_power_off` is required to turn off the machine. Once the mainboard receives the command successfully, it will beep and then proceed to shut down the machine.
Turning the machine on requires two commands. First, the `command_pre_power_on` command is sent to the machine. During normal operation this is usually the first valid message. After that either `command_power_with_cleaning` or `command_power_without_cleaning` must be sent to the machine.
Once both commands have been received, the coffee machine should beep and proceed with the power on sequence.
If the procedure is performed this way, the display will not turn on. Disconnecting and reconnecting power from the display can trick it into turning on (in combination with messages arriving from the mainboard). This is handled by the MOSFET/transistor.

The power commands with and without cleaning can be retrieved by observing multiple starts of the machine. After unplugging and re-plugging the device it will always perform a cleaning cycle. If the machine has already been turned on and it's still warm, a subsequent start will use the `command_pre_power_on` (without cleaning) message.

You can try you commands locally by altering to extending [commands.h](https://github.com/TillFleisch/ESPHome-Philips-Smart-Coffee/blob/main/components/philips_coffee_machine/commands.h).

**Reply by AndrewHath103** [reply]
Edit 2: Fixed it now, believe the third pin on the picoflex is also ground and we had it connected directly across to the mainboard cable. So while pin to was connected to the transistor pin 3 wouldn't so the reboot wouldn't happen. We just disconnected it and now it works properly.


Edit 1: I have since figured out the commands from my last reply which is now quoted, i used a serial port monitor with the `Serial.write(swSer.read());`. It was found to use `byte prePowerOn[] =   {0xD5, 0x55, 0x0A, 0x01, 0x02, 0x00, 0x09, 0x00, 0x00, 0x00, 0x09, 0x15};;` and `byte powerOnTwo[] =   {0xD5, 0x55, 0x01, 0x01, 0x02, 0x00, 0x09, 0x00, 0x00, 0x00, 0x22, 0x20}`. Now only problem i have is getting the display to light up. I do have an npn transistor for the display GND.



> When i press the power button all i ever seem to get is {0xD5, 0x55, 0x00, 0x01, 0x02, 0x00, 0x09, 0x01, 0x00, 0x00, 0x1A, 0x3C} which i believe is the power off command. For some reason I never get anything for powering it on. It may be due to the fact the the code i have doesn't seem to make the machine do anything. If I were to change the loop to just be `Serial.write(swSer.read());` the machine will act like normal but in the serial monitor all i get is squares and occasional letters.
> 
> ```
> > #include <ESP8266WiFi.h>
> #include <SoftwareSerial.h>
> #if defined(ESP8266) && !defined(D5)
> #define D5 (14)
> #define D6 (12)
> #define D7 (13)
> #endif
> SoftwareSerial swSer(D5, D6); // SoftwareSerial to read Display TX
> void setup() {
> Serial.begin(115200); // Initialize serial for debugging
> swSer.begin(115200); // Initialize SoftwareSerial
> // Initialize other components and settings here
> }
> void loop() {
> // Check for available data from the coffee machine
> while (swSer.available() > 0) {
> char receivedByte = swSer.read();
> Serial.print(receivedByte, HEX); // Print received byte in hexadecimal format
> Serial.print(" "); // Add space for readability
> }
> // You can add other logic or processing here if needed
> }
> 
> ```

**Reply by TillFleisch** [reply]
All grounds from the display must be "disconnected", as shown in this schematic:
![image](https://github.com/TillFleisch/ESPHome-Philips-Smart-Coffee/assets/20563797/539cd740-36d2-432f-bc87-3e0132378614)

Does the power on procedure work with and without cleaning?

**Comment by BasJansen** [comment]
Did someone get all this working on a EP1220 yet? Would love to try this on my machine.

**Comment by djratza** [comment]
Hi, I am also have the same device EP1220, does the power on example from Andrew work?

I am new at this, have some knowledge with 433mhz cloning and sending, esp8226 is new to me.

Question i wired directly on the flex cable 5v, ground, tx and rx from the mainboard, is that enough to trigger power on and espresso?

Can i then use the machine with display when not using the ESP?

**Comment by TillFleisch** [comment]
The `EP1220` is currently not supported by this repo.

>  i wired directly on the flex cable 5v, ground, tx and rx from the mainboard, is that enough to trigger power on and espresso?

Yes and No.
This external component has not been designed with 'headless operation' in mind (no connected display). The status sensor might not work and similarly other things may break.
If you also connect the display:
While you *can* start the machine, the display unit will stay turned of. The mosfet/transistor is required to temporarily remove power from the display, which will trick it into turning on.

> Can i then use the machine with display when not using the ESP?

If wired correctly (both the mainboard and display including the required mosfet), you can use the display as usual in addition to control through the ESP.

**Comment by djratza** [comment]
My EP1220 not sure which revison seems to have the same codes as the DEFAULTS in your commands.h.

I was able to catch the off command and it is like in the defaults:

D5:55:00:01:02:00:02:01:00:00:1D:3B

and also the one ending in 11:36

I can see the status sensor reporting back:

[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Idle'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Off'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Idle'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Off'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Idle'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Off'

Not sure why it also displays Off and in the same time also Idle.

I will wire it correctly as per the diagramm today without the Mosfet for now as it is beeing still shipped. Until I get the Mosfet, if I start the CoffeMachine by pressing the Display TouchButton, will the display work? or until the Mosfet is set it will be always powered off, and only usable using the NodeMCU?

Danke fuer deine liebe Antworte

**Comment by djratza** [comment]
Hi, I did the wiring by cutting the ribbon cable, I added your code, put both mainboard and display on debug, I see the code beeing send but that is all. I connected all the unlabeled cables as they were between the mainboard and the display. No transistor for now. i can not power on the machine by button, what could I have done wrong? Putting cables as they where initially withou the nodemcu and the machine powers on

**Comment by djratza** [comment]
Hi, i redid the wiring, now I can start it, but atatus goes to ON then to OFF, I can quickly do a expresso, then it poweres off. Some ideas to twick the status? Display is completly OFF no transistor yet still waiting for shippment, should the display work when starting from the display and not from the ESP?

Thank you, any ideas for status would really help

**Comment by djratza** [comment]
I can confirm that the default commands work for model EP1220, for some reason the wiring did not work on my side, I wired instead the display pin 4 from display to pin 5 from mainboard, and pin 5 to ESP to controll it, I used a relay and created a new button when i power on the relay button is triggered and cut out after 1 second, I do not get the status but I don't care the idea is to wake up with premade coffee and have the display on. Everything works also from the display, thank you for the great project, and sorry for the spam! Have a great Christmas with good coffee.

**Comment by TillFleisch** [comment]
Sorry for the late response.


> should the display work when starting from the display and not from the ESP?

It should, at least it does so on my machine.


Perhaps there is a Hardware difference between the two models, that makes this less compatible, but 
```
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Idle'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Off'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Idle'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Off'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Idle'
[06:36:11][D][text_sensor:064]: 'Status': Sending state 'Off'
```
leads me to believe that this could also be caused by an underlying software issue in this repo. Sadly, debugging without access to one of those machines is difficult.

**Comment by ralenot** [comment]
Hi, I like this project. Is the procedure for EP1220 resolved?
